### ISSUE TITLE
Request: provide css that only applies to classes, not tags

### ISSUE THREAD

**Issue by dpgalloway** [uswds/uswds#2241]
I want to use specific components of the uswds CSS, for example alerts. However, when I link to the uswds.css stylesheet, it overrides almost all of the styles on the site, no matter what order the link tag is placed. This happens because there are several styles in the CSS that apply to tags as a whole, e.g. html {...}, body {...}, so the style is applied no matter what.

I would like to request an alternate CSS file/repo that only applies styles to elements with the specified classes, i.e. only apply styles when an element has a class that starts with "usa-" and leave everything else as-is.

**Comment by thisisdano** [comment]
I've run into this scoping issue myself (adding the government banner to digitalgov.gov) and it hurts. Our base element styling (and our current use of rems) effectively prevents integrating the USWDS into existing sites. I had to do a lot to limit the scope of the USWDS styles to only my component.

I would also hope that this kind of change would be appropriate for our primary css as well (not just a variant), or possibly configured through something like a sass variable. But it may be tricky! (It always is.)

**Comment by maya** [comment]
TODO:
- Document what elements (files) have sizing styles applied to them. 
  - Take a look at normalize and font sizes.
- Think about applying mixins to these elements to either apply to base elements or add'l classes
- Question: should default be applied to base elements or not

**Comment by maya** [comment]
Here's a proposal that gets to this:

1. By default, typography styles are not applied to directly to elements with `$typography-base: false !default;`
2. If users would like typography styles applied to base typography elements, they'll add `$typography-base: true`
3. To use the typography styles, nest the elements inside a `usa-content` element (could also be called `.prose-*` like [Mapbox's Assembly](https://www.mapbox.com/assembly/documentation/#Prose), but we already use `usa-content`)
4. Use the styles as classes with [utility classes](https://v2.designsystem.digital.gov/utilities/font-size-and-family/) e.g. `.font-body-md`, `.font-heading-lg`

```
$typography-base: false !default; // Sets the default typography base styles to false, will be wrapped in usa-content

@mixin typography-base($typography-base) {

  @if $typography-base == false {
    // Opening .usa-content tag
    .usa-content {
  } 
      // Text styles go here
      p {
        @include paragraph-text;
      }  

      h1 {
        @include h1;
      }

      // Additional text styles go here (e.g. headings, lists, etc...)

  @if $typography-base == false {
    } // Closing tag
  }
}

// Class-based paragraph text styling
.usa-p {
  @include paragraph-text;
}

.usa-h1 {
  @include h1;
}
```

Note: I have not tested the mixin yet to see if it actually works, but was shared here as a concept.

**Comment by scottkellum** [comment]
It sounds like there may be a need for you to have the option of compiled CSS without styled elements.

What if you had the regular `uswds.scss` file and a `uswds-components.scss` file for just the components. This way you can link to one or the other depending on your needs.

A lot of the styles in `/elements/` has utility classes alongside element selectors which may be hard to pick out in a simple way like in the button partial. If I were to do this I would have a custom normalize file and include styling for all basic elements alongside that file. This should clean up and consolidate the base styles everything else is build on while not linking to these file(s) would allow for other things to be namespaced.

If you want to style `.usa-button` the same as `button` then you can always use `@extend`

**Comment by maya** [comment]
Got some feedback on what GOV.UK does for their design system:

>I’d say this has been one of the most interesting challenges of the GOV.UK Design System. We wanted complete separation, if a user imported our frontend package into their app it wouldn’t leak out, and their code wouldn’t leak in.
>
>Although this has a huge amount of benefit for the ecosystem as a whole, in research many of our users found needing to add classes to basic HTML elements quite frustrating.
>
>Initially we tried a scoped container where we would style HTML inside a container with a particular class, this was really good for type and some basic components but ended up with some specificity wars with some of our components.
>
>Where we are now is that for (currently only) `<p>` and `<a>` tags we have a global switch `$govuk-global-styles: false !default;`. This lets the user decide if they want us to style their default HTML or not.

**Comment by maya** [comment]
It looks like we'll go with some variation on "Proposal 1" with three options: 
- Scoped styles
- Global element styles
- Only class-based styles, no global element styles and no scoped styles

**Comment by maya** [comment]
Thanks everybody for your thoughtful feedback. This is resolved in #2599 🎉 

Feel free to give it a spin and let us know how it's working for you.
